commons/all/src/main/java/org/infinispan/commons/api/query/Query.java

```diff
@@ -91,46 +92,113 @@ public interface Query<T> extends Iterable<T> {
     *
     * @param paramValues a Map of parameters
     * @return itself
     */
    Query<T> setParameters(Map<String, Object> paramValues);
 
    /**
-    * Returns a {@link CloseableIterator} over the results. Close the iterator when you are done with processing
-    * the results.
+    * Returns a {@link CloseableIterator} over the results. Close the iterator when you are done with processing the
+    * results.
     *
     * @return the results of the query as an iterator.
     */
    @Override
    CloseableIterator<T> iterator();
 
    /**
-    * Returns a {@link CloseableIterator} over the results, including both key and value. Close the iterator when
-    * you are done with processing the results. The query cannot use projections.
+    * Returns a {@link CloseableIterator} over the results, including both key and value. Close the iterator when you
+    * are done with processing the results. The query cannot use projections.
     *
     * @return the results of the query as an iterator.
     */
    <K> CloseableIterator<EntityEntry<K, T>> entryIterator();
 
    /**
-    * Set the timeout for this query. If the query hasn't finished processing before the timeout,
-    * a timeout will be thrown. For queries that use the index, the timeout
-    * is handled on a best effort basis, and the supplied time is rounded to the nearest millisecond.
+    * Set the timeout for this query. If the query hasn't finished processing before the timeout, a timeout will be
+    * thrown. For queries that use the index, the timeout is handled on a best effort basis, and the supplied time is
+    * rounded to the nearest millisecond.
     */
    Query<T> timeout(long timeout, TimeUnit timeUnit);
 
    /**
     * Set the query execution scope
     *
-    * @param local if true, query will be restricted to the data present in the local node, ignoring the other
-    *                  members of the clusters
+    * @param local if true, query will be restricted to the data present in the local node, ignoring the other members
+    *              of the clusters
     */
    Query<T> local(boolean local);
 
    Query<T> scoreRequired(boolean scoreRequired);
 
+   /**
+    * Executes the query (a SELECT statement).
+    * <p>
+    * The method behaves the same as {@link #execute()} but it does not block the invoker thread. Instead, a
+    * {@link CompletionStage} is returned and completed when the statement finishes.
+    * <p>
+    * Not all implementation support non-blocking/async invocation. Use {@link #supportsAsync()} to make sure this
+    * method works as expected.
+    *
+    * @return A {@link CompletionStage} to be completed with the {@link QueryResult}.
+    * @throws UnsupportedOperationException if the implementation does not support non-blocking/async invocation.
+    * @see #execute()
+    * @see #supportsAsync()
+    */
+   default CompletionStage<QueryResult<T>> executeAsync() {
+      throw new UnsupportedOperationException();
```

Comment:
I don't know if it is a good idea of changing the base `Query` class for a case that is supported only for the Remote query using HotRod.

Reply:
`org.infinispan.client.hotrod.impl.query.RemoteQuery` is in a private package and not documented. Also, I hope we will have non-blocking/async implementation for indexed queries at some point.

Reply:
Sorry but I want to touch all the queries for a change that affecting only the client. I don't like the idea of introducing complexity.

Reply:
for this

> I hope we will have non-blocking/async implementation for indexed queries at some point.

I think we should wait for non-blocking Lucene :P

Reply:
I think it is a mistake to wait on the Lucene to do this. At worst case we can offload any internal blocking calls to a blocking thread as we do already for everything. Embedded non indexed queries already support non blocking, but isn't calling the appropriate method today is all.

The API change is rather trivial imo as well as it is just a CompletionStage around the current return values.

Reply:
To be honest I think it would be even better if we had a Flow based return value for the list method as well so that the API isn't a limiting factor to having to keep all results in memory at one time and the user could instead have back pressure aware requests.

Reply:
> Embedded non indexed queries already support non blocking, but isn't calling the appropriate method today is all.

@wburns where?

Having some kind of reactive/stream API would be nice to avoid downloading everything in one go. Keycloak API returns a stream when searching for sessions with some criteria and I'm trying to avoid downloading everything in one go using pagination [1]. The problem with pagination is that I have to sort the results which I assume won't come from free.

[1] https://github.com/pruivo/keycloak/commit/717e1762070e2ed061c8a0dd7453de793ff5b0b2#diff-70c3d04cc474a714e99491df1bab19f3ab6c26c2baf8c094a6ec7dade774fba7R72-R140

Reply:
I can implement "non-blocking" in embedded using `BlockingManager#supplyBlocking(query::execute)`. Would that be ok with you? @fax4ever

Reply:
> > Embedded non indexed queries already support non blocking, but isn't calling the appropriate method today is all.
> 
> @wburns where?

https://github.com/infinispan/infinispan/blob/main/query-core/src/main/java/org/infinispan/query/core/impl/EmbeddedQuery.java#L91

This could be using the ClusterPublisher instead which is fully reactive. It would also help HybridQuery, but I haven't looked into that much.

Reply:
My proposal is to limit this contribution to the remote queries of the HotRod client.

In the meantime we can open a followup issue to make the queries non blocking (or simulate them with an offloading, since Lucene is totally blocking).
In any I prefer doing myself, as maintainer of the query.

Reply:
please leave queries under my maintenance

Reply:
I created https://issues.redhat.com/browse/ISPN-16326